lib/Dialect/mhlo/transforms/mhlo_fusion.cc

```diff
@@ -545,6 +545,40 @@ struct MhloFusionPass : public MhloFusionPassBase<MhloFusionPass> {
       for (Value v : outputs) {
         output_types.push_back(v.getType());
       }
+      //      /-----\
+      //     /       V
+      // A(fused) -- B -- C(fused) -- D(fused)
+      // mlir ops Adjacency List likely above
+      // the B is consumer of fused A, so B need move behind D
+      // because fusion op create at D's location
+      DenseSet<Operation*> fused_set(pattern.begin(), pattern.end());
+      DenseSet<Operation*> consumers_set;
+      SmallVector<Operation*, 4> consumers_vec;
+      auto first_iter = pattern.front()->getIterator();
+      auto last_iter = pattern.back()->getIterator();
+      for (Operation& cur_op : llvm::make_range(first_iter, last_iter)) {
+        // isn't fused op && consumer's op
+        // move this after fusion op
+        if (!fused_set.contains(&cur_op)) {
+          // fused op's consumer or consumer's consumer
+          bool is_consumer = llvm::any_of(
+              cur_op.getOperands(), [&fused_set, &consumers_set](Value v) {
+                auto op = v.getDefiningOp();
+                return fused_set.find(op) != fused_set.end() ||
+                       consumers_set.find(op) != consumers_set.end();
+              });
+          if (is_consumer) {
+            consumers_set.insert(&cur_op);
+            consumers_vec.push_back(&cur_op);
+          }
+        }
+        ++first_iter;
+      }
+      Operation* lastest_op = pattern.back();
+      for (auto op : consumers_vec) {
+        op->moveAfter(lastest_op);
+        lastest_op = op;
+      }
```

Comment:
This works, but can you make sure the difference with the previous version is caught in a test?

Also alternatively it could have been just:
```
 for (Operation *op : llvm::reverse(consumers_vec))
       op->moveAfter(last_iter);
```

Reply:
thanks for review, I think I need to add more test cases